File: src/odom_publisher.cpp
/**
 * @file odom_publisher.cpp
 * @brief Node for publishing odom topic and TF to base_footprint from odom
 * @author Takaki Ueno
 */

// roscpp
#include <ros/ros.h>

// nav_msgs
#include <nav_msgs/Odometry.h>

// geomtery_msgs
#include <geometry_msgs/PoseStamped.h>
#include <geometry_msgs/TwistStamped.h>
#include <geometry_msgs/Quaternion.h>

// tf
#include <tf/transform_datatypes.h>
#include <tf/transform_listener.h>
#include <tf/transform_broadcaster.h>

//! Storage for local position
geometry_msgs::PoseStamped local_pos;

/**
 * @brief Callback function for local position
 * @param msg Incaming msg
 */
void local_pos_cb(const geometry_msgs::PoseStamped::ConstPtr& msg)
{
  local_pos = *msg;
}

//! Storage for local velocity
geometry_msgs::TwistStamped local_vel;

/**
 * @brief Callback for local velocity
 * @param msg Incoming message
 */

void local_vel_cb(const geometry_msgs::TwistStamped::ConstPtr& msg)
{
  local_vel = *msg;
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "odom_publisher");
  ros::NodeHandle nh;
  ros::Rate rate(20);

  // Publisher
  ros::Publisher odom_pub = nh.advertise<nav_msgs::Odometry>("odom", 50);
  // Subscriber
  ros::Subscriber local_pos_sub = nh.subscribe<geometry_msgs::PoseStamped>("mavros/local_position/pose", 50, local_pos_cb);
  ros::Subscriber local_vel_sub = nh.subscribe<geometry_msgs::TwistStamped>("mavros/local_position/velocity", 50, local_vel_cb);

  while(ros::ok())
  {
    ros::spinOnce();

    tf::Quaternion tf_quat = tf::createQuaternionFromYaw(local_pos.pose.orientation.z);

    geometry_msgs::Quaternion odom_quat;
    tf::quaternionTFToMsg(tf_quat, odom_quat);

    nav_msgs::Odometry odom;
    odom.header.stamp = ros::Time::now();
    odom.header.frame_id = "odom";

    odom.pose.pose.position.x = local_pos.pose.position.x;
    odom.pose.pose.position.y = local_pos.pose.position.y;
    odom.pose.pose.position.z = 0.0;
    odom.pose.pose.orientation = odom_quat;

    odom.child_frame_id = "base_link";
    odom.twist.twist.linear.x = local_vel.twist.linear.x;
    odom.twist.twist.linear.y = local_vel.twist.linear.y;
    odom.twist.twist.angular.z = local_vel.twist.angular.z;

    odom_pub.publish(odom);
    rate.sleep();
  }
}

File: script/odom_publisher.py
#!/usr/bin/env python

import rospy
from nav_msgs.msg import Odometry
from geometry_msgs.msg import PoseStamped, TwistStamped, Quaternion
import tf

# Callback for local position
local_pos = PoseStamped()
def local_pos_cb(msg):
    # Need global declaration in order to assign global variable
    global local_pos
    local_pos = msg

# Callback for local velocity
local_vel = TwistStamped()
def local_vel_cb(msg):
    # Need global declaration in order to assign global variable
    global local_vel
    local_vel = msg

def odom_publisher():
    rospy.init_node("odom_publisher")

    odom_pub = rospy.Publisher("odom", Odometry, queue_size=50)

    local_pos_sub = rospy.Subscriber("mavros/local_position/pose",
                                      PoseStamped,
                                      callback=local_pos_cb)
    local_vel_sub = rospy.Subscriber("mavros/local_position/velocity",
                                      TwistStamped,
                                      callback=local_vel_cb)

    rate = rospy.Rate(20)

    while not rospy.is_shutdown():

        pose = local_pos.pose
        twist = local_vel.twist

        # Create quaternion from euler angle of local position
        odom_quat = tf.transformations.quaternion_from_euler(pose.orientation.x,
                                                             pose.orientation.y,
                                                             pose.orientation.z)

        odom = Odometry()
        odom.header.stamp = rospy.Time.now()
        # Global frame is "odom"
        odom.header.frame_id = "odom"

        odom.pose.pose.position.x = pose.position.x
        odom.pose.pose.position.y = pose.position.y
        odom.pose.pose.position.z = 0
        odom.pose.pose.orientation = odom_quat

        # Set robot's base frame as odometry child frame
        odom.child_frame_id = "base_link"
        odom.twist.twist.linear.x = twist.linear.x
        odom.twist.twist.linear.y = twist.linear.y
        odom.twist.twist.angular.z = twist.angular.z

        odom_pub.publish(odom)
        rate.sleep()

# Only run if executed as top level script
if __name__ == "__main__":
    try:
        odom_publisher()
    except rospy.ROSInterruptException:
        pass